### PR TITLE
WD-5038 - fix: Closing panels clears all query params (new solution)

### PR DESCRIPTION
**Pull request by vladimir-cucu**
## Done

- Closing a panel clears only the query params associated with the panel.

## Details

- Fixes: #1502 
https://warthogs.atlassian.net/browse/WD-5038
- Old version of the pull request: https://github.com/canonical/juju-dashboard/pull/1518.

### CODE REVIEW

**Review by huwshimi — COMMENTED** [review]
This is looking good. There's one other place I saw that needs to use the new `handleRemovePanelQueryParams()` function, which is here: https://github.com/canonical/juju-dashboard/blob/88cfd30d7554a02463f5fc2fbdc123f48ba0b971/src/panels/ActionsPanel/CharmActionsPanel.tsx#L196

File: src/panels/utils.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      <tr className="panel__tr" key={th + td}>
+        <th className="panel__th">{th}</th>
+        <td className="panel__td">{td}</td>
+      </tr>
+    );
+  });
+};
+
+// TODO: Add tests for usePanelQueryParams.
+export const usePanelQueryParams = <P extends QueryParams>(
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/panels/utils.tsx]: Oh, my bad 😅 I've relocated the hook in `hooks.ts` and tried to add a few tests for it. Let me know how I can improve them!

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 2]
```diff
@@ -100,7 +105,7 @@ export default function ConfigPanel(): JSX.Element {
   }, [sendAnalytics]);
 
   const onClose = () =>
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/panels/ConfigPanel/ConfigPanel.tsx]: You are right! Thanks for the remark. I've changed it here and the similar issue in `CharmActionsPanel.tsx` as well.

File: src/components/Panel/Panel.tsx [hunk 3]
```diff
@@ -7,81 +7,87 @@ import { forwardRef, useId } from "react";
...
+      onRemovePanelQueryParams,
+      ...props
+    }: Props,
     ref
   ) => {
-    const [, setPanelQs] = useQueryParams<{ panel: string | null }>({
-      panel: null,
-    });
+    // TODO: Make onRemovePanelQueryParams a required prop,
+    // remove handleRemovePanelQueryParams and modify tests.
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/components/Panel/Panel.tsx]: Removed both TODO comments, made the optional param `onRemovePanelQueryParams` in `Panel` and `PanelHeader` be required, and fixed/added a few tests for `Panel` and `PanelHeader`.

File: src/components/Panel/Panel.test.tsx [hunk 4]
```diff
@@ -1,129 +1,158 @@
...
+      url: "/foo?panel=share-model&externalParam=externalValue",
+    });
+  };
+
+  it("should render successfully", () => {
+    const {
+      result: { container },
+    } = mockRenderComponent({});
+
+    expect(container).toMatchSnapshot();
```
Comment (huwshimi) [inline src/components/Panel/Panel.test.tsx:45]: This is fine, but we've generally been moving away from snapshots now as they're a lot less useful with RTL as it doesn't support shallow rendering.

**Review by huwshimi — APPROVED** [review]
This is working really nicely. I've left a couple of comments about things that don't need to be fixed in this PR, but rather things to consider for the future.

File: src/components/Panel/Panel.test.tsx (hunk 4 above)
Comment (vladimir-cucu) [reply src/components/Panel/Panel.test.tsx]: Fixed it in the latest commit and merged the PR.

**Pull request merged**